### PR TITLE
Planner uses IndexScan for partially-covering indexes

### PR DESCRIPTION
**Pull request by alex-cash**
Follow-up to #334.

Appended to the index rows' keys the corresponding PK values from the originating table, so the row can be looked up when there's an index scan using a partially-covering index. The executor gets the values for columns covered by the index, then gets the remaining column values from the originating table for each row.

An index is partially-covering when some, but not all of the columns queried are in the index, and where the "left-most" index column is queried. Index PKs are comprised of `[shardId][tableId][col0][col1]...[colN][tablePKCol0][tablePKCol1]...[tablePKColN]`, so to get the index key prefix, `col0` has to be in the query.

### CODE REVIEW

**Comment by purplefox** [comment]
Alex, quick question: With this PR do we support indexes with more than one column?

**Review by purplefox — COMMENTED** [review]
As a general comment on the approach taken here. Right now the approach seems to be to take as many columns as possible from the index and then take any missing columns from the table. This could be simplified a lot by either a) taking all the rows from the index if it covers or b) just take all the columns from the table. A lot of the complexity in the code seems to be in dealing with some of the columns coming from the index and some from the table, but that seems unnecessary. That also means changes to common decodeKeyCols() etc wouldn't be necessary. Happy to discuss more :)

File: pull/exec/index_reader.go [hunk 1]
```diff
@@ -31,18 +30,22 @@ func NewPullIndexReader(tableInfo *common.TableInfo,
 	includedCols []int, // Col indexes in the table that are being returned
 	storage cluster.Cluster,
 	shardID uint64,
-	scanRanges []*ScanRange,
-	covers bool) (*PullIndexReader, error) {
+	scanRanges []*ScanRange) (*PullIndexReader, error) {
 
 	// Calculate the types of the results
 	var resultColTypes []common.ColumnType
 	for _, colIndex := range includedCols {
```
Comment (purplefox) [inline pull/exec/index_reader.go:37]: Instead of iterating through includedClose twice, could be done once

**Comment by alex-cash** [comment]
@purplefox: The current state of this PR is how we left it after our pairing session. There are a few outstanding items to address on follow-up PRs, which I'll create issues for:
- column ordering is different between table creation and the physical plan
- composite ranges aren't supported (so `select * from transactions where customer_id = 100 and col2 = 'abc'` will attempt to use an index scan on `index1` and fail)
- optimizations re: defining included PK cols, queried non-index cols, if the index covers, etc. should occur when index reader is instantiated

**Comment by purplefox** [comment]
I've fixed the issue with schema being different from output columns and also fixed non deterministic tests. Going to merge this PR now and follow up with the other issues

**Pull request merged**